{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import tarfile\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['boto3', 'botocore', 'urllib3', 'requests', 'wheel', 'certifi', 'idna', 'typing-extensions', 'charset-normalizer', 'pip', 'setuptools', 'python-dateutil', 's3transfer', 'packaging', 'aiobotocore', 'pyyaml', 'six', 's3fs', 'fsspec', 'numpy', 'cryptography', 'grpcio-status', 'google-api-core', 'cffi', 'pycparser', 'pypular', 'pandas', 'importlib-metadata', 'pyasn1', 'rsa', 'zipp', 'attrs', 'click', 'pydantic', 'protobuf', 'jmespath', 'platformdirs', 'pytz', 'jinja2', 'colorama', 'markupsafe', 'pyjwt', 'awscli', 'tomli', 'wrapt', 'google-auth', 'googleapis-common-protos', 'filelock', 'cachetools', 'requests-oauthlib', 'oauthlib', 'pluggy', 'virtualenv', 'pyarrow', 'docutils', 'jsonschema', 'pyasn1-modules', 'pytest', 'exceptiongroup', 'aiohttp', 'pyparsing', 'sqlalchemy', 'scipy', 'isodate', 'multidict', 'psutil', 'pyopenssl', 'yarl', 'iniconfig', 'decorator', 'soupsieve', 'pygments', 'tzdata', 'async-timeout', 'beautifulsoup4', 'frozenlist', 'aiosignal', 'tqdm', 'grpcio', 'pillow', 'requests-toolbelt', 'greenlet', 'pydantic-core', 'openpyxl', 'et-xmlfile', 'lxml', 'werkzeug', 'pynacl', 'tomlkit', 'proto-plus', 'importlib-resources', 'asn1crypto', 'azure-core', 'deprecated', 'distlib', 'websocket-client', 'annotated-types', 'coverage', 'google-cloud-storage', 'more-itertools', 'h11', 'msgpack', 'sniffio', 'pexpect', 'ptyprocess', 'referencing', 'scikit-learn', 'rpds-py', 'google-cloud-core', 'flask', 'chardet', 'anyio', 'backoff', 'psycopg2-binary', 'jsonschema-specifications', 'google-resumable-media', 'matplotlib', 'python-dotenv', 'tenacity', 'msal', 'poetry-plugin-export', 'py', 'tabulate', 'bcrypt', 'httpcore', 'wcwidth', 'httpx', 'poetry-core', 'rich', 'pathspec', 'mypy-extensions', 'regex', 'dill', 'paramiko', 'gitpython', 'threadpoolctl', 'google-auth-oauthlib', 'itsdangerous', 'snowflake-connector-python', 'kiwisolver', 'ruamel-yaml', 'sortedcontainers', 'keyring', 'cycler', 'smmap', 'google-crc32c', 'shellingham', 'joblib', 'networkx', 'gitdb', 'markdown-it-py', 'mdurl', 'pymysql', 'azure-storage-blob', 'portalocker', 'py4j', 'fonttools', 'build', 'sqlparse', 'secretstorage', 'jaraco-classes', 'jeepney', 'google-api-python-client', 'awswrangler', 'fastjsonschema', 'grpcio-tools', 'tzlocal', 'marshmallow', 'nest-asyncio', 'msal-extensions', 'pyproject-hooks', 'google-cloud-bigquery', 'pkginfo', 'rapidfuzz', 'typedload', 'prompt-toolkit', 'trove-classifiers', 'xmltodict', 'pytest-cov', 'cython', 'dnspython', 'pendulum', 'distro', 'scramp', 'azure-identity', 'cloudpickle', 'uritemplate', 'httplib2', 'toml', 'google-auth-httplib2', 'pyrsistent', 'defusedxml', 'docker', 'traitlets', 'ruamel-yaml-clib', 'ply', 'mccabe', 'pytzdata', 'future', 'termcolor', 'requests-aws4auth', 'ipython', 'cachecontrol', 'blinker', 'babel', 'markdown', 'contourpy', 'tornado', 'pygithub', 'redis', 'poetry', 'gunicorn', 'msrest', 'alembic', 'retry', 'azure-common', 'jsonpointer', 'jedi', 'huggingface-hub', 'datadog', 'pycryptodomex', 'redshift-connector', 'pycodestyle', 'isort', 'pg8000', 'webencodings', 'crashtest', 'toolz', 'parso', 'matplotlib-inline', 'dulwich', 'jsonpath-ng', 'tb-nightly', 'cleo', 'fastapi', 'prometheus-client', 'grpc-google-iam-v1', 'types-requests', 'black', 'installer', 'kubernetes', 'multiprocess', 'python-json-logger', 'pkgutil-resolve-name', 'google-pasta', 'pyzmq', 'pycryptodome', 'sentry-sdk', 'jsonpatch', 'arrow', 'mako', 'typer', 'opensearch-py', 'jupyter-core', 'transformers', 'websockets', 'progressbar2', 'torch', 'pyspark', 'asttokens', 'jupyter-client', 'executing', 'pymongo', 'typing-inspect', 'aioitertools', 'tensorboard', 'setuptools-scm', 'opentelemetry-api', 'stack-data', 'oscrypto', 'pure-eval', 'elasticsearch', 'argcomplete', 'google-cloud-pubsub', 'asgiref', 'pytest-runner', 'python-utils', 'ipykernel', 'orjson', 'aenum', 'uvicorn', 'types-python-dateutil', 'jupyter-server', 'bs4', 'pyflakes', 'starlette', 'nbconvert', 'absl-py', 'debugpy', 'shapely', 'loguru', 'smart-open', 'tokenizers', 'nbformat', 'requests-file', 'google-cloud-secret-manager', 'pyodbc', 'structlog', 'sagemaker', 'snowflake-sqlalchemy', 'notebook', 'flake8', 'rfc3339-validator', 'apache-airflow', 'altair', 'adal', 'xlrd', 'mysql-connector-python', 'monotonic', 'mistune', 'tinycss2', 'humanfriendly', 'pylint', 'zope-interface', 'bleach', 'astroid', 'h5py', 'pysocks', 'sentencepiece', 'comm', 'opentelemetry-sdk', 'orbax-checkpoint', 'tensorflow', 'imageio', 'asynctest', 'jupyterlab-server', 'typeguard', 'openai', 'appdirs', 'that-depends', 'cattrs', 'pbr', 'nltk', 'great-expectations', 'xlsxwriter', 'mpmath', 'sympy', 'gremlinpython', 'send2trash', 'google-cloud-bigquery-storage', 'jupyterlab', 'argon2-cffi', 'opentelemetry-semantic-conventions', 'oauth2client', 'nbclient', 'xgboost', 'argon2-cffi-bindings', 'tensorstore', 'gast', 'simplejson', 'setproctitle', 'text-unidecode', 'flatbuffers', 'pytest-xdist', 'slack-sdk', 'overrides', 'json5', 'execnet', 'notebook-shim', 'croniter', 'ujson', 'watchdog', 'db-dtypes', 'pathos', 'dataclasses-json', 'analytics-python', 'fast-version', 'schema', 'pandocfilters', 'email-validator', 'pox', 'ppft', 'python-slugify', 'google-cloud-aiplatform', 'jupyterlab-pygments', 'widgetsnbextension', 'ipywidgets', 'keras', 'google-cloud-resource-manager', 'pkce', 'evergreen-py', 'lz4', 'nodeenv', 'smdebug-rulesconfig', 'webcolors', 'aiofiles', 'jupyter-events', 'pre-commit', 'mypy', 'db-contrib-tool', 'identify', 'seaborn', 'fqdn', 'rfc3986-validator', 'tensorflow-estimator', 'isoduration', 'uri-template', 'jupyterlab-widgets', 'entrypoints', 'terminado', 'google-cloud-appengine-logging', 'tensorboard-data-server', 'safetensors', 'html5lib', 'hvac', 'numba', 'lazy-object-proxy', 'semver', 'statsmodels', 'llvmlite', 'jupyter-server-terminals', 'azure-mgmt-core', 'cfgv', 'selenium', 'cached-property', 'async-lru', 'azure-storage-file-datalake', 'pytest-mock', 'databricks-sql-connector', 'docker-pycreds', 'mock', 'databricks-cli', 'click-man', 'pandas-gbq', 'opt-einsum', 'jupyter-lsp', 'thrift', 'apache-airflow-providers-common-sql', 'psycopg2', 'patsy', 'lockfile', 'makefun', 'msrestazure', 'aws-requests-auth', 'ndg-httpsclient', 'opentelemetry-proto', 'wsproto', 'httptools', 'backports-zoneinfo', 'responses', 'pickleshare', 'antlr4-python3-runtime', 'wandb', 'semantic-version', 'uvloop', 'astunparse', 'python-multipart', 'django', 'click-plugins', 'backcall', 'opencv-python', 'google-cloud-logging', 'sshtunnel', 'plotly', 'tensorflow-io-gcs-filesystem', 'pydata-google-auth', 'langchain-core', 'dataclasses', 'ordered-set', 'gsutil', 'gcsfs', 'tox', 'azure-keyvault-secrets', 'watchfiles', 'argparse', 'scikit-image', 'aws-sam-translator', 'prettytable', 'faker', 'sphinx', 'confluent-kafka', 'spacy', 'kafka-python', 'pydeequ', 'contextlib2', 'inflection', 'coloredlogs', 'jaraco-functools', 'torchvision', 'zeep', 'retrying', 'unidecode', 'rfc3986', 'docstring-parser', 'opentelemetry-exporter-otlp-proto-common', 'libcst', 'pymssql', 'azure-datalake-store', 'xxhash', 'imbalanced-learn', 'tldextract', 'graphql-core', 'opentelemetry-exporter-otlp-proto-http', 'libclang', 'hatchling', 'jaraco-context', 'mlflow', 'snowballstemmer', 'trio', 'colorlog', 'delta-spark', 'azure-mgmt-resource', 'azure-nspkg', 'deprecation', 'google-cloud-firestore', 'outcome', 'pipenv', 'google-cloud-audit-log', 'opentelemetry-exporter-otlp-proto-grpc', 'pywavelets', 'catalogue', 'pysam', 'pytest-metadata', 'wasabi', 'triton', 'tensorflow-serving-api', 'querystring-parser', 'deepdiff', 'cymem', 'jsonpickle', 'tblib', 'srsly', 'murmurhash', 'fastavro', 'edgegrid-python', 'preshed', 'gspread', 'ruff', 'opencensus', 'opencensus-context', 'docopt', 'pybind11', 'aniso8601', 'azure-storage-common', 'datasets', 'graphviz', 'uv', 'cinemagoer', 'imdbpy', 'configparser', 'more-executors', 'fastpurge', 'thinc', 'opencensus-ext-azure', 'backports-tarfile', 'pytest-asyncio', 'ml-dtypes', 'flask-appbuilder', 'google-cloud-kms', 'trio-websocket', 'nvidia-nvjitlink-cu12', 'datetime', 'google-cloud-dlp', 'google-cloud-vision', 'langchain-community', 'opentelemetry-exporter-otlp', 'simple-salesforce', 'nvidia-nccl-cu12', 'ninja', 'imagesize', 'apache-airflow-providers-snowflake', 'brotli', 'google-cloud-monitoring', 'google-cloud-dataproc', 'pysftp', 'blis', 'watchtower', 'azure-mgmt-storage', 'alabaster', 'ecdsa', 'apache-airflow-providers-databricks', 'google-cloud-spanner', 'sphinxcontrib-serializinghtml', 'gevent', 'freezegun', 'mdit-py-plugins', 'looker-sdk', 'chex', 'pytest-html', 'azure-graphrbac', 'tiktoken', 'nvidia-cudnn-cu12', 'omegaconf', 'mypy-boto3-rds', 'pywin32', 'sphinxcontrib-htmlhelp', 'nvidia-cublas-cu12', 'nvidia-cuda-nvrtc-cu12', 'typing', 'sphinxcontrib-qthelp', 'sphinxcontrib-devhelp', 'boto', 'sphinxcontrib-applehelp', 'zstandard', 'apache-airflow-providers-google', 'zope-event', 'kombu', 'jira', 'azure-keyvault', 'nvidia-cuda-runtime-cu12', 'google-cloud-bigtable', 'nvidia-cuda-cupti-cu12', 'sphinxcontrib-jsmath', 'nvidia-cusparse-cu12', 'nvidia-cusolver-cu12', 'nvidia-cufft-cu12', 'python-gnupg', 'azure-cosmos', 'python-daemon', 'inject', 'amqp', 'applicationinsights', 'nvidia-curand-cu12', 'google-ads', 'google-cloud-tasks', 'fastapi-cli', 'cfn-lint', 'celery', 'pyathena', 'types-pyyaml', 'nvidia-nvtx-cu12', 'langchain', 'bitarray', 'google-cloud-container', 'pycrypto', 'dbt-core', 'vine', 'python-magic', 'google-cloud-datacatalog', 'dask', 'billiard', 'authlib', 'apispec', 'optax', 'pycountry', 'azure-keyvault-keys', 'mashumaro', 'pyproj', 'flask-cors', 'djangorestframework', 'google-cloud-bigquery-datatransfer', 'apache-airflow-providers-ssh', 'azure-mgmt-containerregistry', 'google-cloud-language', 'google-cloud-videointelligence', 'invoke', 'google-cloud-build', 'azure-storage-queue', 'pathlib2', 'google-cloud-workflows', 'google-cloud-redis', 'texttable', 'accelerate', 'pypdf2', 'azure-mgmt-authorization', 'google-cloud-automl', 'azure-mgmt-network', 'moto', 'python-jose', 'flit-core', 'stevedore', 'apache-airflow-providers-mysql', 'protobuf3-to-dict', 'pip-tools', 'azure-mgmt-keyvault', 'google-cloud-os-login', 'ijson', 'einops', 'google-cloud-memcache', 'langcodes', 'azure-data-tables', 'gradio', 'pytimeparse', 'hpack', 'tensorflow-text', 'cramjam', 'azure-mgmt-compute', 'apache-airflow-providers-http', 'spacy-legacy', 'google-cloud-translate', 'cligj', 'google-cloud-dataplex', 'humanize', 'crcmod', 'azure-cli', 'jpype1', 'opentelemetry-instrumentation', 'hyperframe', 'mypy-boto3-s3', 'h2', 'langsmith', 'apache-beam', 'datadog-api-client', 'hypothesis', 'pyproject-api', 'ratelimit', 'spacy-loggers', 'lightgbm', 'enum34', 'holidays', 'click-didyoumean', 'sqlalchemy-bigquery', 'grpcio-gcp', 'sh', 'graphene', 'graphql-relay', 'mysqlclient', 'cmake', 'google-cloud-speech', 'python-http-client', 'click-repl', 'apscheduler', 'google-cloud-dataproc-metastore', 'confection', 'azure-mgmt-cosmosdb', 'cssselect', 'validators', 'google-cloud-orchestration-airflow', 'avro', 'agate', 'google-cloud-texttospeech', 'avro-python3', 'dateparser', 'ipython-genutils', 'jsondiff', 'shap', 'aws-psycopg2', 'parsedatetime', 'fuzzywuzzy', 'boto3-stubs', 'elastic-transport', 'google-cloud-dataform', 'office365-rest-python-client', 'azure-mgmt-nspkg', 'tifffile', 'fastparquet', 'flask-sqlalchemy', 'fire', 'azure-mgmt-msi', 'polars', 'readme-renderer', 'junit-xml', 'fasteners', 'botocore-stubs', 'time-machine', 'futures', 'nvidia-cudnn-cu11', 'nvidia-cublas-cu11', 'pytest-timeout', 'sendgrid', 'gcloud-aio-storage', 'azure-mgmt-datalake-nspkg', 'azure-mgmt-containerinstance', 'tensorboard-plugin-wit', 'azure-mgmt-recoveryservices', 'gcloud-aio-auth', 'fabric', 'types-protobuf', 'azure-mgmt-monitor', 'ddtrace', 'jax', 'azure-mgmt-redis', 'pyspnego', 'nvidia-cuda-runtime-cu11', 'sklearn', 'nvidia-cuda-nvrtc-cu11', 'typed-ast', 'azure-servicebus', 'azure-batch', 'scp', 'knack', 'azure-mgmt-sql', 'azure-mgmt-signalr', 'jaydebeapi', 'gcloud-aio-bigquery', 'nose', 'azure-mgmt-datalake-store', 'yamllint', 'nh3', 'azure-mgmt-web', 'astor', 'azure-mgmt-containerservice', 'onnxruntime', 'twine', 'parameterized', 'unicodecsv', 'phonenumbers', 'azure-mgmt-dns', 'pipx', 'azure-mgmt-servicebus', 'py-cpuinfo', 'google-cloud-compute', 'pydot', 'pyserial', 'flask-wtf', 'geopandas', 'slicer', 'types-setuptools', 'azure-mgmt-eventhub', 'resolvelib', 'lazy-loader', 'azure-mgmt-rdbms', 'azure-mgmt-batch', 'pyarrow-hotfix', 'marisa-trie', 'fiona', 'azure-mgmt-loganalytics', 'slackclient', 'databricks-sdk', 'azure-mgmt-cdn', 'azure-mgmt-iothub', 'pytest-rerunfailures', 'levenshtein', 'flask-login', 'kfp', 'azure-mgmt-trafficmanager', 'azure-mgmt-recoveryservicesbackup', 'azure-mgmt-marketplaceordering', 'azure-mgmt-devtestlabs', 'pyroaring', 'tf-keras', 'opentelemetry-util-http', 'pathy', 'language-data', 'azure-mgmt-search', 'azure-mgmt-cognitiveservices', 'azure-mgmt-managementgroups', 'parse', 'ddsketch', 'azure-mgmt-applicationinsights', 'azure-mgmt-eventgrid', 'azure-mgmt-servicefabric', 'azure-mgmt-datafactory', 'azure-mgmt-advisor', 'azure-mgmt-billing', 'azure-mgmt-media', 'azure-mgmt-iothubprovisioningservices', 'universal-pathlib', 'azure-mgmt-batchai', 'azure-mgmt-policyinsights', 'types-urllib3', 'azure-mgmt-iotcentral', 'azure-mgmt-datamigration', 'types-awscrt', 'azure-mgmt-maps', 'python-gitlab', 'azure-storage-file-share', 'mypy-boto3-appflow', 'lark', 'azure-cli-core', 'yapf', 'wtforms', 'azure-eventhub', 'tensorflow-metadata', 'python-docx', 'inflect', 'sqlalchemy-utils', 'leather', 'azure-appconfiguration', 'iso8601', 'apache-airflow-providers-fab', 'onnx', 'bytecode', 'funcsigs', 'geoip2', 'types-pytz', 'python-levenshtein', 'opencv-python-headless', 'dbt-extractor', 'starkbank-ecdsa', 'tensorflow-hub', 'connexion', 'django-cors-headers', 'partd', 'cloudpathlib', 'netaddr', 'azure-mgmt-datalake-analytics', 'qtconsole', 'hyperlink', 'torchaudio', 'setuptools-rust', 'pytorch-lightning', 'cron-descriptor', 'locket', 'geographiclib', 'keras-preprocessing', 'envier', 'timm', 'dm-tree', 'pyotp', 'geopy', 'xarray', 'ansible', 'limits', 'pika', 'azure-mgmt-apimanagement', 'diskcache', 'types-s3transfer', 'dacite', 'psycopg', 'cachelib', 'uamqp', 'gensim', 'ftfy', 'maxminddb', 'azure-loganalytics', 'marshmallow-enum', 'apache-airflow-providers-sqlite', 'cloud-tpu-client', 'qtpy', 'apache-airflow-providers-ftp', 'pyhcl', 'statsd', 'azure-mgmt-consumption', 'gradio-client', 'azure-mgmt-relay', 'rich-argparse', 'torchmetrics', 'langchain-text-splitters', 'flask-caching', 'junitparser', 'apache-airflow-providers-common-io', 'prometheus-flask-exporter', 'requests-mock', 'pydantic-settings', 'azure-multiapi-storage', 'azure-mgmt-reservations', 'dbt-semantic-interfaces', 'apache-airflow-providers-smtp', 'javaproperties', 'google-re2', 'pyee', 'keras-applications', 'mlflow-skinny', 'awscrt', 'lightning-utilities', 'pypdf', 'pydub', 'marshmallow-sqlalchemy', 'google-cloud', 'pathlib', 'jupyter-console', 'flask-limiter', 'configargparse', 'ansible-core', 'autopep8', 'linkify-it-py', 'reportlab', 'faiss-cpu', 'opentelemetry-instrumentation-asgi', 'flask-jwt-extended', 'orderedmultidict', 'azure-cli-telemetry', 'azure-synapse-artifacts', 'strenum', 'requests-ntlm', 'mergedeep', 'aws-xray-sdk', 'uc-micro-py', 'pytest-forked', 'azure-keyvault-certificates', 'convertdate', 'userpath', 'apache-airflow-providers-imap', 'natsort', 'jupyter', 'cerberus', 'apache-airflow-providers-amazon', 'commonmark', 'azure-mgmt-hdinsight', 'azure-mgmt-privatedns', 'spark-nlp', 'bracex', 'asyncio', 'azure-synapse-spark', 'azure-mgmt-security', 'jaxlib', 'azure-mgmt-kusto', 'azure-mgmt-managedservices', 'swagger-ui-bundle', 'elasticsearch-dsl', 'ipaddress', 'pyaml', 'azure-mgmt-netapp', 'azure-devops', 'constructs', 'azure-mgmt-redhatopenshift', 'azure-keyvault-administration', 'azure-mgmt-synapse', 'kfp-pipeline-spec', 'azure-mgmt-imagebuilder', 'django-filter', 'azure-mgmt-servicelinker', 'azure-mgmt-appconfiguration', 'azure-mgmt-sqlvirtualmachine', 'opentelemetry-instrumentation-fastapi', 'azure-mgmt-servicefabricmanagedclusters', 'azure-mgmt-botservice', 'openapi-spec-validator', 'datefinder', 'azure-synapse-accesscontrol', 'nbclassic', 'polyfactory', 'azure-mgmt-databoxedge', 'azure-synapse-managedprivateendpoints', 'magicattr', 'checkov', 'incremental', 'azure-mgmt-extendedlocation', 'jsonlines', 'boltons', 'ua-parser', 'tensorflow-datasets', 'ldap3', 'mmh3', 'twisted', 'google-cloud-dataflow-client', 'langdetect', 'stringcase', 'sentence-transformers', 'flax', 'editables', 'pydash', 'bandit', 'ray', 'portpicker', 'azure-kusto-data', 'binaryornot', 'jellyfish', 'xlwt', 'cookiecutter', 'bokeh', 'cog', 'emoji', 'h3', 'azure-cosmosdb-table', 'oracledb', 'logbook', 'sphinx-rtd-theme', 'korean-lunar-calendar', 'distributed', 'pymeeus', 'mkdocs-material', 'aws-lambda-powertools', 'cx-oracle', 'methodtools', 'webdriver-manager', 'pytest-split', 'dbt-snowflake', 'urllib3-secure-extra', 'json-merge-patch', 'streamlit', 'azure-cosmosdb-nspkg', 'pathlib-abc', 'pgpy', 'immutabledict', 'pdfminer-six', 'frozendict', 'tensorflow-probability', 'etils', 'pyelftools', 'cloudevents', 'tensorflow-io', 'passlib', 'mypy-boto3-redshift-data', 'python-decouple', 'pandas-stubs', 'automat', 'constantly', 'flask-session', 'minimal-snowplow-tracker', 'atomicwrites', 'async-generator', 'promise', 'pyperclip', 'boostedblob', 'azureml-core', 'numexpr', 'python-socketio', 'plac', 'oldest-supported-numpy', 'python-engineio', 'grpcio-health-checking', 'wirerope', 'apache-airflow-providers-slack', 'dpath', 'opentelemetry-instrumentation-requests', 'bidict', 'cssselect2', 'moreorless', 'pyrfc3339', 'asyncpg', 'eth-account', 'functions-framework', 'service-identity', 'weasel', 'uuid', 'plotnine', 'opentelemetry-instrumentation-wsgi', 'trailrunner', 'pytest-messenger', 'stdlibs', 'unittest-xml-reporting', 'flask-babel', 'usort', 'ufmt', 'wcmatch', 'django-extensions', 'sqlalchemy-redshift', 'blessed', 'scandir', 'jdcal', 'firebase-admin', 'feedparser', 'rdflib', 'prison', 'configupdater', 'pytest-randomly', 'parse-type', 'array-record', 'kazoo', 'azure-mgmt-deploymentmanager', 'dynamodb-json', 'configobj', 'kfp-server-api', 'tomli-w', 'psycopg-binary', 'xyzservices', 'twilio', 'google', 'yappi', 'django-storages', 'addict', 'optree', 'playwright', 'bottle', 'marshmallow-oneofschema', 'google-cloud-datastore', 'factory-boy', 'sarif-om', 'tableauserverclient', 'jschema-to-python', 'apache-airflow-providers-cncf-kubernetes', 'editdistance', 'distrax', 'hijri-converter', 'fs', 'namex', 'pytz-deprecation-shim', 'sqlalchemy-spanner', 'mkdocs', 'sphinxcontrib-jquery', 'user-agents', 'aiosqlite', 'w3lib', 'python3-openid', 'cmdstanpy', 'teradatasql', 'pytest-django', 'google-cloud-run', 'openapi-schema-validator', 'hydra-core', 'mypy-boto3-glue', 'qrcode', 'simple-websocket', 'google-cloud-storage-transfer', 'newrelic', 'datasketch', 'google-cloud-batch', 'pymdown-extensions', 'ultralytics', 'pycares', 'pydocstyle', 'expiringdict', 'fasttext-wheel', 'aiodns', 'opentelemetry-instrumentation-flask', 'scikit-build-core', 'ipdb', 'hdfs', 'stripe', 'pymsteams', 'adlfs', 'mizani', 'pypandoc', 'opentelemetry-instrumentation-dbapi', 'click-option-group', 'shortuuid', 'hiredis', 'python-box', 'python-editor', 'opentelemetry-instrumentation-urllib', 'kaleido', 'azure-kusto-ingest', 'pyhumps', 'gql', 'diff-cover', 'azure-storage-file', 'pydeck', 'ghp-import', 'pypika', 'jsii', 'evaluate', 'tfds-nightly', 'pyhocon', 'lru-dict', 'pyyaml-env-tag', 'yq', 'pyquery', 'dynaconf', 'trino', 'multimethod', 'pymupdf', 'google-apitools', 'types-redis', 'dash', 'opentelemetry-instrumentation-urllib3', 'sqlalchemy-jsonfield', 'gviz-api', 'librosa', 'tensorboard-plugin-profile', 'netifaces', 'grpcio-reflection', 'telethon', 'pycocotools', 'findspark', 'jsonref', 'hologram', 'pyhive', 'comtypes', 'django-redis', 'launchdarkly-server-sdk', 'rasterio', 'terminaltables', 'affine', 'speechbrain', 'filetype', 'robotframework', 'python-consul', 'hyperpyyaml', 'clu', 'clickclick', 'databricks-api', 'strictyaml', 'eth-rlp', 'pyinstaller-hooks-contrib', 'pynumdiff', 'publication', 'pooch', 'opencensus-ext-logging', 'toposort', 'enum-compat', 'pypng', 'scapy', 'lit', 'concurrent-log-handler', 'pyglove', 'marshmallow-dataclass', 'pep517', 'altgraph', 'sseclient-py', 'whitenoise', 'msgspec', 'netcdf4', 'python-jenkins', 'azure-eventgrid', 'snuggs', 'voluptuous', 'diffusers', 'duckdb', 'seqio', 'django-debug-toolbar', 'tokenize-rt', 'soundfile', 'dbt-postgres', 'dm-env', 'sudachidict-core', 'cftime', 'pyphen', 'pyinstaller', 'opentelemetry-instrumentation-psycopg2', 'mypy-boto3-secretsmanager', 'mypy-boto3-dynamodb', 'azure-monitor-opentelemetry-exporter', 'hatch-vcs', 'rich-click', 'peewee', 'hyperopt', 'waitress', 'rlax', 'cytoolz', 'opentelemetry-instrumentation-django', 'django-environ', 'geventhttpclient', 'pyudev', 'furl', 'pypiwin32', 'sqlglot', 'eth-hash', 'ffmpy', 'pmdarima', 'aiohttp-retry', 'google-cloud-pubsublite', 'pywin32-ctypes', 'arviz', 'minio', 'webob', 'chevron', 'koalas', 'ciso8601', 'opencv-contrib-python', 'av', 'influxdb', 'sudachipy', 'pyxlsb', 'backports-functools-lru-cache', 'prophet', 'pathvalidate', 'flake8-bugbear', 'bitstring', 'mypy-boto3-lambda', 'mypy-protobuf', 'pynamodb', 'dash-core-components', 'dash-html-components', 'allure-python-commons', 'dash-table', 'schedule', 'markdown2', 'somnium', 'fixedint', 'ffmpeg-python', 'timezonefinder', 'atlassian-python-api', 'beartype', 'sacremoses', 'catboost', 'multipledispatch', 'graphframes', 'pylev', 'strip-hints', 'smbprotocol', 'injector', 'json-log-formatter', 'mypy-boto3-sqs', 'langchain-openai', 'flower', 'optuna', 'thriftpy2', 'cleanco', 'singledispatch', 'apache-airflow-providers-postgres', 'elementpath', 'pywinauto', 'pytest-env', 'aws-cdk-lib', 'tensorboardx', 'readchar', 'hatch-fancy-pypi-readme', 'mkdocs-material-extensions', 'zict', 'fpdf', 'pytesseract', 'htpasswd', 'thop', 'djangorestframework-simplejwt', 'salesforce-bulk', 'cfn-flip', 'sgmllib3k', 'python-snappy', 'singer-sdk', 'xtermcolor', 'colour', 'pystan', 'yt-dlp', 'eth-utils', 'backports-weakref', 'sql-metadata', 'glom', 'recordlinkage', 'face', 'xmlschema', 'snowflake-snowpark-python', 'open-clip-torch', 'azure', 'flask-openid', 'polling', 'aws-cdk-asset-awscli-v1', 'albumentations', 'docker-compose', 'pathable', 'pyppeteer', 'appnope', 'fakeredis', 'requests-aws-sign', 'pipdeptree', 'o365', 'imageio-ffmpeg', 'geomet', 'iopath', 'signalfx', 'wordcloud', 'nox', 'google-cloud-recommendations-ai', 'posthog', 'redis-py-cluster', 'deltalake', 'trimesh', 'thrift-sasl', 'uritools', 'azure-mgmt-subscription', 'jaconv', 'drf-yasg', 'audioread', 'types-toml', 'jwcrypto', 'filterpy', 'py-spy', 'ec2-metadata', 'num2words', 'visions', 'hexbytes', 'virtualenv-clone', 'requests-unixsocket', 'multi-key-dict', 'azure-monitor-opentelemetry', 'pkgconfig', 'py-partiql-parser', 'flask-restful', 'pyusb', 'types-cachetools', 'python-rapidjson', 'coreapi', 'paho-mqtt', 'pygsheets', 'colorful', 'types-six', 'python-crontab', 'itypes', 'pytest-repeat', 'pymupdfb', 'eth-typing', 'django-timezone-field', 'weasyprint', 'parsimonious', 'azure-core-tracing-opentelemetry', 'tablib', 'sacrebleu', 'cerberus-python-client', 'pygeohash', 'hatch', 'aioboto3', 'ephem', 'click-default-group', 'requests-futures', 'cssutils', 'meson', 'nvidia-ml-py', 'pyfiglet', 'olefile', 'mysql-connector', 'opentelemetry-resource-detector-azure', 'pybloom-live', 'pynvml', 'backports-tempfile', 'peft', 'appium-python-client', 'boa-str', 'kubernetes-asyncio', 'wget', 'mypy-boto3-cloudformation', 'allure-pytest', 'langchain-google-vertexai', 'stanio', 'clickhouse-driver', 'pygit2', 'sqlfluff', 'types-pyopenssl', 'motor', 'nvidia-cuda-cupti-cu11', 'deepmerge', 'dbutils', 'hjson', 'pandera', 'packageurl-python', 'htmlmin', 'drf-spectacular', 'nvidia-cusolver-cu11', 'mutagen', 'nvidia-nccl-cu11', 'nvidia-cusparse-cu11', 'python-nvd3', 'ansible-compat', 'boxsdk', 'myst-parser', 'flashtext', 'dominate', 'nvidia-cufft-cu11', 'nvidia-curand-cu11', 'pytest-custom-exit-code']\n"
     ]
    }
   ],
   "source": [
    "def get_top_pypi_packages(url='https://hugovk.github.io/top-pypi-packages/top-pypi-packages-30-days.json'):\n",
    "    response = requests.get(url)\n",
    "    response.raise_for_status() \n",
    "    \n",
    "    data = response.json()\n",
    "    packages = [package['project'] for package in data['rows'][:1500]]\n",
    "\n",
    "    return packages\n",
    "\n",
    "top_packages = get_top_pypi_packages()\n",
    "print(top_packages)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting boto3\n",
      "  Using cached boto3-1.34.136.tar.gz (108 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/boto3/boto3-1.34.136.tar.gz\n",
      "Successfully downloaded boto3\n",
      "Downloaded boto3 into /mnt/volume_nyc1_01/benignPyPI/boto3\n",
      "Progress: 1/1500\n",
      "Collecting botocore\n",
      "  Using cached botocore-1.34.136.tar.gz (12.6 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/botocore/botocore-1.34.136.tar.gz\n",
      "Successfully downloaded botocore\n",
      "Downloaded botocore into /mnt/volume_nyc1_01/benignPyPI/botocore\n",
      "Progress: 2/1500\n",
      "Collecting urllib3\n",
      "  Using cached urllib3-2.2.2.tar.gz (292 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/urllib3/urllib3-2.2.2.tar.gz\n",
      "Successfully downloaded urllib3\n",
      "Downloaded urllib3 into /mnt/volume_nyc1_01/benignPyPI/urllib3\n",
      "Progress: 3/1500\n",
      "Collecting requests\n",
      "  Using cached requests-2.32.3.tar.gz (131 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/requests/requests-2.32.3.tar.gz\n",
      "Successfully downloaded requests\n",
      "Downloaded requests into /mnt/volume_nyc1_01/benignPyPI/requests\n",
      "Progress: 4/1500\n",
      "Collecting wheel\n",
      "  Using cached wheel-0.43.0.tar.gz (99 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/wheel/wheel-0.43.0.tar.gz\n",
      "Successfully downloaded wheel\n",
      "Downloaded wheel into /mnt/volume_nyc1_01/benignPyPI/wheel\n",
      "Progress: 5/1500\n",
      "Collecting certifi\n",
      "  Using cached certifi-2024.6.2.tar.gz (165 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/certifi/certifi-2024.6.2.tar.gz\n",
      "Successfully downloaded certifi\n",
      "Downloaded certifi into /mnt/volume_nyc1_01/benignPyPI/certifi\n",
      "Progress: 6/1500\n",
      "Collecting idna\n",
      "  Using cached idna-3.7.tar.gz (189 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/idna/idna-3.7.tar.gz\n",
      "Successfully downloaded idna\n",
      "Downloaded idna into /mnt/volume_nyc1_01/benignPyPI/idna\n",
      "Progress: 7/1500\n",
      "Collecting typing-extensions\n",
      "  Using cached typing_extensions-4.12.2.tar.gz (85 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/typing-extensions/typing_extensions-4.12.2.tar.gz\n",
      "Successfully downloaded typing-extensions\n",
      "Downloaded typing-extensions into /mnt/volume_nyc1_01/benignPyPI/typing-extensions\n",
      "Progress: 8/1500\n",
      "Collecting charset-normalizer\n",
      "  Using cached charset-normalizer-3.3.2.tar.gz (104 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/charset-normalizer/charset-normalizer-3.3.2.tar.gz\n",
      "Successfully downloaded charset-normalizer\n",
      "Downloaded charset-normalizer into /mnt/volume_nyc1_01/benignPyPI/charset-normalizer\n",
      "Progress: 9/1500\n",
      "Collecting pip\n",
      "  Using cached pip-24.1.1.tar.gz (1.9 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/pip/pip-24.1.1.tar.gz\n",
      "Successfully downloaded pip\n",
      "Downloaded pip into /mnt/volume_nyc1_01/benignPyPI/pip\n",
      "Progress: 10/1500\n",
      "Collecting setuptools\n",
      "  Using cached setuptools-70.1.1.tar.gz (2.3 MB)\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/setuptools/setuptools-70.1.1.tar.gz\n",
      "Successfully downloaded setuptools\n",
      "Downloaded setuptools into /mnt/volume_nyc1_01/benignPyPI/setuptools\n",
      "Progress: 11/1500\n",
      "Collecting python-dateutil\n",
      "  Using cached python-dateutil-2.9.0.post0.tar.gz (342 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/python-dateutil/python-dateutil-2.9.0.post0.tar.gz\n",
      "Successfully downloaded python-dateutil\n",
      "Downloaded python-dateutil into /mnt/volume_nyc1_01/benignPyPI/python-dateutil\n",
      "Progress: 12/1500\n",
      "Collecting s3transfer\n",
      "  Using cached s3transfer-0.10.2.tar.gz (144 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/s3transfer/s3transfer-0.10.2.tar.gz\n",
      "Successfully downloaded s3transfer\n",
      "Downloaded s3transfer into /mnt/volume_nyc1_01/benignPyPI/s3transfer\n",
      "Progress: 13/1500\n",
      "Collecting packaging\n",
      "  Using cached packaging-24.1.tar.gz (148 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/packaging/packaging-24.1.tar.gz\n",
      "Successfully downloaded packaging\n",
      "Downloaded packaging into /mnt/volume_nyc1_01/benignPyPI/packaging\n",
      "Progress: 14/1500\n",
      "Collecting aiobotocore\n",
      "  Using cached aiobotocore-2.13.1.tar.gz (104 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/aiobotocore/aiobotocore-2.13.1.tar.gz\n",
      "Successfully downloaded aiobotocore\n",
      "Downloaded aiobotocore into /mnt/volume_nyc1_01/benignPyPI/aiobotocore\n",
      "Progress: 15/1500\n",
      "Collecting pyyaml\n",
      "  Using cached PyYAML-6.0.1.tar.gz (125 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/pyyaml/PyYAML-6.0.1.tar.gz\n",
      "Successfully downloaded pyyaml\n",
      "Downloaded pyyaml into /mnt/volume_nyc1_01/benignPyPI/pyyaml\n",
      "Progress: 16/1500\n",
      "Collecting six\n",
      "  Using cached six-1.16.0.tar.gz (34 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/six/six-1.16.0.tar.gz\n",
      "Successfully downloaded six\n",
      "Downloaded six into /mnt/volume_nyc1_01/benignPyPI/six\n",
      "Progress: 17/1500\n",
      "Collecting s3fs\n",
      "  Using cached s3fs-2024.6.1.tar.gz (74 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/s3fs/s3fs-2024.6.1.tar.gz\n",
      "Successfully downloaded s3fs\n",
      "Downloaded s3fs into /mnt/volume_nyc1_01/benignPyPI/s3fs\n",
      "Progress: 18/1500\n",
      "Collecting fsspec\n",
      "  Using cached fsspec-2024.6.1.tar.gz (284 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/fsspec/fsspec-2024.6.1.tar.gz\n",
      "Successfully downloaded fsspec\n",
      "Downloaded fsspec into /mnt/volume_nyc1_01/benignPyPI/fsspec\n",
      "Progress: 19/1500\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.0.0.tar.gz (18.3 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'error'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[91 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Collecting meson-python>=0.15.0\n",
      "  \u001b[31m   \u001b[0m   Using cached meson_python-0.16.0.tar.gz (82 kB)\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Installing backend dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing backend dependencies: finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install backend dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m╰─>\u001b[0m \u001b[31m[68 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Collecting ninja>=1.8.2\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Using cached ninja-1.11.1.1.tar.gz (132 kB)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Collecting patchelf>=0.11.0\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Using cached patchelf-0.17.2.1.tar.gz (167 kB)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Building wheels for collected packages: ninja, patchelf\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Building wheel for ninja (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Building wheel for ninja (pyproject.toml): finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m×\u001b[0m \u001b[32mBuilding wheel for ninja \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mpyproject.toml\u001b[0m\u001b[1;32m)\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m╰─>\u001b[0m \u001b[31m[15 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m /tmp/pip-build-env-rpr7v8vb/overlay/local/lib/python3.12/dist-packages/setuptools_scm/git.py:312: UserWarning: git archive did not support describe output\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   warnings.warn(\"git archive did not support describe output\")\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m /tmp/pip-build-env-rpr7v8vb/overlay/local/lib/python3.12/dist-packages/setuptools_scm/git.py:331: UserWarning: unprocessed git archival found (no export subst applied)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   warnings.warn(\"unprocessed git archival found (no export subst applied)\")\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   File \"/tmp/pip-build-env-rpr7v8vb/overlay/local/lib/python3.12/dist-packages/skbuild/setuptools_wrap.py\", line 643, in setup\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m     cmkr = cmaker.CMaker(cmake_executable)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   File \"/tmp/pip-build-env-rpr7v8vb/overlay/local/lib/python3.12/dist-packages/skbuild/cmaker.py\", line 145, in __init__\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m     self.cmake_version = get_cmake_version(self.cmake_executable)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   File \"/tmp/pip-build-env-rpr7v8vb/overlay/local/lib/python3.12/dist-packages/skbuild/cmaker.py\", line 102, in get_cmake_version\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m     raise SKBuildError(msg) from err\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Problem with the CMake installation, aborting build. CMake executable is cmake\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m  ERROR: Failed building wheel for ninja\u001b[0m\u001b[31m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[0m  Building wheel for patchelf (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   Building wheel for patchelf (pyproject.toml): finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m×\u001b[0m \u001b[32mBuilding wheel for patchelf \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mpyproject.toml\u001b[0m\u001b[1;32m)\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m╰─>\u001b[0m \u001b[31m[11 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   File \"/tmp/pip-build-env-833s5jnz/overlay/local/lib/python3.12/dist-packages/skbuild/setuptools_wrap.py\", line 643, in setup\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m     cmkr = cmaker.CMaker(cmake_executable)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   File \"/tmp/pip-build-env-833s5jnz/overlay/local/lib/python3.12/dist-packages/skbuild/cmaker.py\", line 145, in __init__\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m     self.cmake_version = get_cmake_version(self.cmake_executable)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   File \"/tmp/pip-build-env-833s5jnz/overlay/local/lib/python3.12/dist-packages/skbuild/cmaker.py\", line 102, in get_cmake_version\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m     raise SKBuildError(msg) from err\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Problem with the CMake installation, aborting build. CMake executable is cmake\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m   \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m  ERROR: Failed building wheel for patchelf\u001b[0m\u001b[31m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[0mFailed to build ninja patchelf\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31mERROR: Could not build wheels for ninja, patchelf, which is required to install pyproject.toml-based projects\u001b[0m\u001b[31m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install backend dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \u001b[31m╰─>\u001b[0m See above for output.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download numpy: Command '['/bin/python3', '-m', 'pip', 'download', 'numpy', '--no-binary=:all:', '--no-deps', '--dest', '/mnt/volume_nyc1_01/benignPyPI/numpy']' returned non-zero exit status 1.\n",
      "Retrying in 1 seconds...\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.0.0.tar.gz (18.3 MB)\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py\", line 180, in exc_logging_wrapper\n",
      "    status = run_func(*args)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/req_command.py\", line 245, in wrapper\n",
      "    return func(self, options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/commands/download.py\", line 132, in run\n",
      "    requirement_set = resolver.resolve(reqs, check_supported_wheels=True)\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/resolver.py\", line 95, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "                            ^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 546, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 397, in resolve\n",
      "    self._add_to_criteria(self.state.criteria, r, parent=None)\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 173, in _add_to_criteria\n",
      "    if not criterion.candidates:\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/structs.py\", line 156, in __bool__\n",
      "    return bool(self._sequence)\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 155, in __bool__\n",
      "    return any(self)\n",
      "           ^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 143, in <genexpr>\n",
      "    return (c for c in iterator if id(c) not in self._incompatible_ids)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 47, in _iter_built\n",
      "    candidate = func()\n",
      "                ^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 182, in _make_candidate_from_link\n",
      "    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(\n",
      "                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 228, in _make_base_candidate_from_link\n",
      "    self._link_candidate_cache[link] = LinkCandidate(\n",
      "                                       ^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 290, in __init__\n",
      "    super().__init__(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 156, in __init__\n",
      "    self.dist = self._prepare()\n",
      "                ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 222, in _prepare\n",
      "    dist = self._prepare_distribution()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 301, in _prepare_distribution\n",
      "    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 525, in prepare_linked_requirement\n",
      "    return self._prepare_linked_requirement(req, parallel_builds)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 640, in _prepare_linked_requirement\n",
      "    dist = _get_prepared_distribution(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 71, in _get_prepared_distribution\n",
      "    abstract_dist.prepare_distribution_metadata(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py\", line 67, in prepare_distribution_metadata\n",
      "    self.req.prepare_metadata()\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py\", line 579, in prepare_metadata\n",
      "    self.metadata_directory = generate_metadata(\n",
      "                              ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/build/metadata.py\", line 35, in generate_metadata\n",
      "    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/utils/misc.py\", line 766, in prepare_metadata_for_build_wheel\n",
      "    return super().prepare_metadata_for_build_wheel(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 186, in prepare_metadata_for_build_wheel\n",
      "    return self._call_hook('prepare_metadata_for_build_wheel', {\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 321, in _call_hook\n",
      "    raise BackendUnavailable(data.get('traceback', ''))\n",
      "pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 77, in _build_backend\n",
      "    obj = import_module(mod_path)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3.12/importlib/__init__.py\", line 90, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1324, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'mesonpy'\n",
      "\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download numpy: Command '['/bin/python3', '-m', 'pip', 'download', 'numpy', '--no-binary=:all:', '--no-deps', '--no-build-isolation', '--dest', '/mnt/volume_nyc1_01/benignPyPI/numpy']' returned non-zero exit status 2.\n",
      "Retrying in 2 seconds...\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.0.0.tar.gz (18.3 MB)\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py\", line 180, in exc_logging_wrapper\n",
      "    status = run_func(*args)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/req_command.py\", line 245, in wrapper\n",
      "    return func(self, options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/commands/download.py\", line 132, in run\n",
      "    requirement_set = resolver.resolve(reqs, check_supported_wheels=True)\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/resolver.py\", line 95, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "                            ^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 546, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 397, in resolve\n",
      "    self._add_to_criteria(self.state.criteria, r, parent=None)\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 173, in _add_to_criteria\n",
      "    if not criterion.candidates:\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/structs.py\", line 156, in __bool__\n",
      "    return bool(self._sequence)\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 155, in __bool__\n",
      "    return any(self)\n",
      "           ^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 143, in <genexpr>\n",
      "    return (c for c in iterator if id(c) not in self._incompatible_ids)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 47, in _iter_built\n",
      "    candidate = func()\n",
      "                ^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 182, in _make_candidate_from_link\n",
      "    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(\n",
      "                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 228, in _make_base_candidate_from_link\n",
      "    self._link_candidate_cache[link] = LinkCandidate(\n",
      "                                       ^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 290, in __init__\n",
      "    super().__init__(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 156, in __init__\n",
      "    self.dist = self._prepare()\n",
      "                ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 222, in _prepare\n",
      "    dist = self._prepare_distribution()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 301, in _prepare_distribution\n",
      "    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 525, in prepare_linked_requirement\n",
      "    return self._prepare_linked_requirement(req, parallel_builds)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 640, in _prepare_linked_requirement\n",
      "    dist = _get_prepared_distribution(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 71, in _get_prepared_distribution\n",
      "    abstract_dist.prepare_distribution_metadata(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py\", line 67, in prepare_distribution_metadata\n",
      "    self.req.prepare_metadata()\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py\", line 579, in prepare_metadata\n",
      "    self.metadata_directory = generate_metadata(\n",
      "                              ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/build/metadata.py\", line 35, in generate_metadata\n",
      "    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/utils/misc.py\", line 766, in prepare_metadata_for_build_wheel\n",
      "    return super().prepare_metadata_for_build_wheel(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 186, in prepare_metadata_for_build_wheel\n",
      "    return self._call_hook('prepare_metadata_for_build_wheel', {\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 321, in _call_hook\n",
      "    raise BackendUnavailable(data.get('traceback', ''))\n",
      "pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 77, in _build_backend\n",
      "    obj = import_module(mod_path)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3.12/importlib/__init__.py\", line 90, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1324, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'mesonpy'\n",
      "\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download numpy: Command '['/bin/python3', '-m', 'pip', 'download', 'numpy', '--no-binary=:all:', '--no-deps', '--no-build-isolation', '--dest', '/mnt/volume_nyc1_01/benignPyPI/numpy']' returned non-zero exit status 2.\n",
      "Retrying in 4 seconds...\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.0.0.tar.gz (18.3 MB)\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py\", line 180, in exc_logging_wrapper\n",
      "    status = run_func(*args)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/req_command.py\", line 245, in wrapper\n",
      "    return func(self, options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/commands/download.py\", line 132, in run\n",
      "    requirement_set = resolver.resolve(reqs, check_supported_wheels=True)\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/resolver.py\", line 95, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "                            ^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 546, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 397, in resolve\n",
      "    self._add_to_criteria(self.state.criteria, r, parent=None)\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 173, in _add_to_criteria\n",
      "    if not criterion.candidates:\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/structs.py\", line 156, in __bool__\n",
      "    return bool(self._sequence)\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 155, in __bool__\n",
      "    return any(self)\n",
      "           ^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 143, in <genexpr>\n",
      "    return (c for c in iterator if id(c) not in self._incompatible_ids)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 47, in _iter_built\n",
      "    candidate = func()\n",
      "                ^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 182, in _make_candidate_from_link\n",
      "    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(\n",
      "                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 228, in _make_base_candidate_from_link\n",
      "    self._link_candidate_cache[link] = LinkCandidate(\n",
      "                                       ^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 290, in __init__\n",
      "    super().__init__(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 156, in __init__\n",
      "    self.dist = self._prepare()\n",
      "                ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 222, in _prepare\n",
      "    dist = self._prepare_distribution()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 301, in _prepare_distribution\n",
      "    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 525, in prepare_linked_requirement\n",
      "    return self._prepare_linked_requirement(req, parallel_builds)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 640, in _prepare_linked_requirement\n",
      "    dist = _get_prepared_distribution(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 71, in _get_prepared_distribution\n",
      "    abstract_dist.prepare_distribution_metadata(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py\", line 67, in prepare_distribution_metadata\n",
      "    self.req.prepare_metadata()\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py\", line 579, in prepare_metadata\n",
      "    self.metadata_directory = generate_metadata(\n",
      "                              ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/build/metadata.py\", line 35, in generate_metadata\n",
      "    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/utils/misc.py\", line 766, in prepare_metadata_for_build_wheel\n",
      "    return super().prepare_metadata_for_build_wheel(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 186, in prepare_metadata_for_build_wheel\n",
      "    return self._call_hook('prepare_metadata_for_build_wheel', {\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 321, in _call_hook\n",
      "    raise BackendUnavailable(data.get('traceback', ''))\n",
      "pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 77, in _build_backend\n",
      "    obj = import_module(mod_path)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3.12/importlib/__init__.py\", line 90, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1324, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'mesonpy'\n",
      "\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download numpy: Command '['/bin/python3', '-m', 'pip', 'download', 'numpy', '--no-binary=:all:', '--no-deps', '--no-build-isolation', '--dest', '/mnt/volume_nyc1_01/benignPyPI/numpy']' returned non-zero exit status 2.\n",
      "Retrying in 8 seconds...\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.0.0.tar.gz (18.3 MB)\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py\", line 180, in exc_logging_wrapper\n",
      "    status = run_func(*args)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/req_command.py\", line 245, in wrapper\n",
      "    return func(self, options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/commands/download.py\", line 132, in run\n",
      "    requirement_set = resolver.resolve(reqs, check_supported_wheels=True)\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/resolver.py\", line 95, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "                            ^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 546, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 397, in resolve\n",
      "    self._add_to_criteria(self.state.criteria, r, parent=None)\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 173, in _add_to_criteria\n",
      "    if not criterion.candidates:\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/structs.py\", line 156, in __bool__\n",
      "    return bool(self._sequence)\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 155, in __bool__\n",
      "    return any(self)\n",
      "           ^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 143, in <genexpr>\n",
      "    return (c for c in iterator if id(c) not in self._incompatible_ids)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 47, in _iter_built\n",
      "    candidate = func()\n",
      "                ^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 182, in _make_candidate_from_link\n",
      "    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(\n",
      "                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 228, in _make_base_candidate_from_link\n",
      "    self._link_candidate_cache[link] = LinkCandidate(\n",
      "                                       ^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 290, in __init__\n",
      "    super().__init__(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 156, in __init__\n",
      "    self.dist = self._prepare()\n",
      "                ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 222, in _prepare\n",
      "    dist = self._prepare_distribution()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 301, in _prepare_distribution\n",
      "    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 525, in prepare_linked_requirement\n",
      "    return self._prepare_linked_requirement(req, parallel_builds)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 640, in _prepare_linked_requirement\n",
      "    dist = _get_prepared_distribution(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 71, in _get_prepared_distribution\n",
      "    abstract_dist.prepare_distribution_metadata(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py\", line 67, in prepare_distribution_metadata\n",
      "    self.req.prepare_metadata()\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py\", line 579, in prepare_metadata\n",
      "    self.metadata_directory = generate_metadata(\n",
      "                              ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/build/metadata.py\", line 35, in generate_metadata\n",
      "    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/utils/misc.py\", line 766, in prepare_metadata_for_build_wheel\n",
      "    return super().prepare_metadata_for_build_wheel(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 186, in prepare_metadata_for_build_wheel\n",
      "    return self._call_hook('prepare_metadata_for_build_wheel', {\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 321, in _call_hook\n",
      "    raise BackendUnavailable(data.get('traceback', ''))\n",
      "pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 77, in _build_backend\n",
      "    obj = import_module(mod_path)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3.12/importlib/__init__.py\", line 90, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1324, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'mesonpy'\n",
      "\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download numpy: Command '['/bin/python3', '-m', 'pip', 'download', 'numpy', '--no-binary=:all:', '--no-deps', '--no-build-isolation', '--dest', '/mnt/volume_nyc1_01/benignPyPI/numpy']' returned non-zero exit status 2.\n",
      "Retrying in 16 seconds...\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.0.0.tar.gz (18.3 MB)\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py\", line 180, in exc_logging_wrapper\n",
      "    status = run_func(*args)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/cli/req_command.py\", line 245, in wrapper\n",
      "    return func(self, options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/commands/download.py\", line 132, in run\n",
      "    requirement_set = resolver.resolve(reqs, check_supported_wheels=True)\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/resolver.py\", line 95, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "                            ^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 546, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 397, in resolve\n",
      "    self._add_to_criteria(self.state.criteria, r, parent=None)\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/resolvers.py\", line 173, in _add_to_criteria\n",
      "    if not criterion.candidates:\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/resolvelib/structs.py\", line 156, in __bool__\n",
      "    return bool(self._sequence)\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 155, in __bool__\n",
      "    return any(self)\n",
      "           ^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 143, in <genexpr>\n",
      "    return (c for c in iterator if id(c) not in self._incompatible_ids)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/found_candidates.py\", line 47, in _iter_built\n",
      "    candidate = func()\n",
      "                ^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 182, in _make_candidate_from_link\n",
      "    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(\n",
      "                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/factory.py\", line 228, in _make_base_candidate_from_link\n",
      "    self._link_candidate_cache[link] = LinkCandidate(\n",
      "                                       ^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 290, in __init__\n",
      "    super().__init__(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 156, in __init__\n",
      "    self.dist = self._prepare()\n",
      "                ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 222, in _prepare\n",
      "    dist = self._prepare_distribution()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/resolution/resolvelib/candidates.py\", line 301, in _prepare_distribution\n",
      "    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 525, in prepare_linked_requirement\n",
      "    return self._prepare_linked_requirement(req, parallel_builds)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 640, in _prepare_linked_requirement\n",
      "    dist = _get_prepared_distribution(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py\", line 71, in _get_prepared_distribution\n",
      "    abstract_dist.prepare_distribution_metadata(\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py\", line 67, in prepare_distribution_metadata\n",
      "    self.req.prepare_metadata()\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py\", line 579, in prepare_metadata\n",
      "    self.metadata_directory = generate_metadata(\n",
      "                              ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/operations/build/metadata.py\", line 35, in generate_metadata\n",
      "    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_internal/utils/misc.py\", line 766, in prepare_metadata_for_build_wheel\n",
      "    return super().prepare_metadata_for_build_wheel(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 186, in prepare_metadata_for_build_wheel\n",
      "    return self._call_hook('prepare_metadata_for_build_wheel', {\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_impl.py\", line 321, in _call_hook\n",
      "    raise BackendUnavailable(data.get('traceback', ''))\n",
      "pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 77, in _build_backend\n",
      "    obj = import_module(mod_path)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3.12/importlib/__init__.py\", line 90, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1324, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'mesonpy'\n",
      "\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download numpy: Command '['/bin/python3', '-m', 'pip', 'download', 'numpy', '--no-binary=:all:', '--no-deps', '--no-build-isolation', '--dest', '/mnt/volume_nyc1_01/benignPyPI/numpy']' returned non-zero exit status 2.\n",
      "Max retry attempts reached for numpy. Skipping...\n",
      "Collecting cryptography\n",
      "  Using cached cryptography-42.0.8.tar.gz (671 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'error'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[72 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Collecting setuptools>=61.0.0\n",
      "  \u001b[31m   \u001b[0m   Using cached setuptools-70.1.1-py3-none-any.whl\n",
      "  \u001b[31m   \u001b[0m Collecting wheel\n",
      "  \u001b[31m   \u001b[0m   Using cached wheel-0.43.0-py3-none-any.whl\n",
      "  \u001b[31m   \u001b[0m Collecting cffi>=1.12\n",
      "  \u001b[31m   \u001b[0m   Using cached cffi-1.16.0.tar.gz (512 kB)\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m Collecting setuptools-rust>=1.7.0\n",
      "  \u001b[31m   \u001b[0m   Using cached setuptools_rust-1.9.0-py3-none-any.whl\n",
      "  \u001b[31m   \u001b[0m Collecting pycparser (from cffi>=1.12)\n",
      "  \u001b[31m   \u001b[0m   Using cached pycparser-2.22-py3-none-any.whl\n",
      "  \u001b[31m   \u001b[0m Collecting semantic-version<3,>=2.8.2 (from setuptools-rust>=1.7.0)\n",
      "  \u001b[31m   \u001b[0m   Using cached semantic_version-2.10.0-py2.py3-none-any.whl\n",
      "  \u001b[31m   \u001b[0m Building wheels for collected packages: cffi\n",
      "  \u001b[31m   \u001b[0m   Building wheel for cffi (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   Building wheel for cffi (pyproject.toml): finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[31m×\u001b[0m \u001b[32mBuilding wheel for cffi \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mpyproject.toml\u001b[0m\u001b[1;32m)\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m╰─>\u001b[0m \u001b[31m[39 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m running bdist_wheel\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m running build\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m running build_py\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m creating build\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-312\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m creating build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/model.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/_imp_emulation.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/_shimmed_dist_utils.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/vengine_cpy.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/commontypes.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/pkgconfig.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/ffiplatform.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/verifier.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/vengine_gen.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/error.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/api.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/lock.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/backend_ctypes.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/recompiler.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/__init__.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/setuptools_ext.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/cffi_opcode.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/cparser.py -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/_cffi_include.h -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/parse_c_type.h -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/_embedding.h -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m copying src/cffi/_cffi_errors.h -> build/lib.linux-x86_64-cpython-312/cffi\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m running build_ext\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m building '_cffi_backend' extension\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-312\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-312/src\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m creating build/temp.linux-x86_64-cpython-312/src/c\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m x86_64-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O2 -Wall -fPIC -DFFI_BUILDING=1 -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/include/ffi -I/usr/include/libffi -I/usr/include/python3.12 -c src/c/_cffi_backend.c -o build/temp.linux-x86_64-cpython-312/src/c/_cffi_backend.o\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m src/c/_cffi_backend.c:15:10: fatal error: ffi.h: No such file or directory\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m    15 | #include <ffi.h>\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m       |          ^~~~~~~\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m compilation terminated.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m error: command '/bin/x86_64-linux-gnu-gcc' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m  ERROR: Failed building wheel for cffi\u001b[0m\u001b[31m\n",
      "  \u001b[31m   \u001b[0m \u001b[0mFailed to build cffi\n",
      "  \u001b[31m   \u001b[0m \u001b[31mERROR: Could not build wheels for cffi, which is required to install pyproject.toml-based projects\u001b[0m\u001b[31m\n",
      "  \u001b[31m   \u001b[0m \u001b[0m\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download cryptography: Command '['/bin/python3', '-m', 'pip', 'download', 'cryptography', '--no-binary=:all:', '--no-deps', '--dest', '/mnt/volume_nyc1_01/benignPyPI/cryptography']' returned non-zero exit status 1.\n",
      "Retrying in 1 seconds...\n",
      "Collecting cryptography\n",
      "  Using cached cryptography-42.0.8.tar.gz (671 kB)\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Saved /mnt/volume_nyc1_01/benignPyPI/cryptography/cryptography-42.0.8.tar.gz\n",
      "Successfully downloaded cryptography\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 7\u001b[0m, in \u001b[0;36mdl_packages\u001b[0;34m(packages, target_directory)\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m----> 7\u001b[0m     \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcheck_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecutable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m-m\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpip\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdownload\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpackage\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--no-binary=:all:\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--no-deps\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--dest\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpackage_dir\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDownloaded \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpackage\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m into \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpackage_dir\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.12/subprocess.py:413\u001b[0m, in \u001b[0;36mcheck_call\u001b[0;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    412\u001b[0m         cmd \u001b[38;5;241m=\u001b[39m popenargs[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m--> 413\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(retcode, cmd)\n\u001b[1;32m    414\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '['/bin/python3', '-m', 'pip', 'download', 'cryptography', '--no-binary=:all:', '--no-deps', '--dest', '/mnt/volume_nyc1_01/benignPyPI/cryptography']' returned non-zero exit status 1.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 35\u001b[0m\n\u001b[1;32m     31\u001b[0m top_packages \u001b[38;5;241m=\u001b[39m get_top_pypi_packages()\n\u001b[1;32m     33\u001b[0m download_directory \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/mnt/volume_nyc1_01/benignPyPI/\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m---> 35\u001b[0m \u001b[43mdl_packages\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtop_packages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdownload_directory\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[15], line 20\u001b[0m, in \u001b[0;36mdl_packages\u001b[0;34m(packages, target_directory)\u001b[0m\n\u001b[1;32m     18\u001b[0m time\u001b[38;5;241m.\u001b[39msleep(wait_time)\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 20\u001b[0m     \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcheck_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecutable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m-m\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpip\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdownload\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpackage\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--no-binary=:all:\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--no-deps\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--no-build-isolation\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--dest\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpackage_dir\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     21\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDownloaded \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpackage\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m into \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpackage_dir\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     22\u001b[0m     count \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "File \u001b[0;32m/usr/lib/python3.12/subprocess.py:408\u001b[0m, in \u001b[0;36mcheck_call\u001b[0;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    398\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcheck_call\u001b[39m(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    399\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run command with arguments.  Wait for command to complete.  If\u001b[39;00m\n\u001b[1;32m    400\u001b[0m \u001b[38;5;124;03m    the exit code was zero then return, otherwise raise\u001b[39;00m\n\u001b[1;32m    401\u001b[0m \u001b[38;5;124;03m    CalledProcessError.  The CalledProcessError object will have the\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    406\u001b[0m \u001b[38;5;124;03m    check_call([\"ls\", \"-l\"])\u001b[39;00m\n\u001b[1;32m    407\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 408\u001b[0m     retcode \u001b[38;5;241m=\u001b[39m \u001b[43mcall\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpopenargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    409\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m retcode:\n\u001b[1;32m    410\u001b[0m         cmd \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124margs\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.12/subprocess.py:391\u001b[0m, in \u001b[0;36mcall\u001b[0;34m(timeout, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    389\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m p:\n\u001b[1;32m    390\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 391\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    392\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m:  \u001b[38;5;66;03m# Including KeyboardInterrupt, wait handled that.\u001b[39;00m\n\u001b[1;32m    393\u001b[0m         p\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m/usr/lib/python3.12/subprocess.py:1264\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1262\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m _time() \u001b[38;5;241m+\u001b[39m timeout\n\u001b[1;32m   1263\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1264\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1265\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1266\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1267\u001b[0m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1268\u001b[0m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1269\u001b[0m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1270\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/lib/python3.12/subprocess.py:2053\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   2051\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   2052\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 2053\u001b[0m (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2054\u001b[0m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   2055\u001b[0m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   2056\u001b[0m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   2057\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pid \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid:\n",
      "File \u001b[0;32m/usr/lib/python3.12/subprocess.py:2011\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   2009\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   2010\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 2011\u001b[0m     (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2012\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[1;32m   2013\u001b[0m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   2014\u001b[0m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   2015\u001b[0m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   2016\u001b[0m     pid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def dl_packages(packages, target_directory):\n",
    "    count = 0\n",
    "    for package in packages:\n",
    "        package_dir = os.path.join(target_directory, package)\n",
    "        os.makedirs(package_dir, exist_ok=True)\n",
    "        retries = 3\n",
    "        delay = 1\n",
    "        while retries > 0:\n",
    "            try:\n",
    "                subprocess.check_call([sys.executable, \"-m\", \"pip\", \"download\", package, \"--no-binary=:all:\", \"--no-deps\", \"--dest\", package_dir])\n",
    "                print(f\"Downloaded {package} into {package_dir}\")\n",
    "                count += 1\n",
    "                print(f\"Progress: {count}/{len(packages)}\")\n",
    "                break  # Break out of the retry loop if download is successful\n",
    "            except subprocess.CalledProcessError as e:\n",
    "                print(f\"Failed to download {package}: {e}\")\n",
    "                retries -= 1\n",
    "                if retries == 0:\n",
    "                    print(f\"Max retries reached for {package}. Skipping.\")\n",
    "                else:\n",
    "                    print(f\"Retrying in {delay} seconds...\")\n",
    "                    time.sleep(delay)\n",
    "                    delay *= 2  # Exponential backoff\n",
    "                    continue\n",
    "\n",
    "top_packages = get_top_pypi_packages()\n",
    "\n",
    "download_directory = '/mnt/volume_nyc1_01/benignPyPI/'\n",
    "\n",
    "dl_packages(top_packages, download_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "repo_dir = \"/mnt/volume_nyc1_01/benignPyPI/\"\n",
    "#Find all tar.gz files in the repository\n",
    "tar_files = []\n",
    "for root, dirs, files in os.walk(repo_dir):\n",
    "    for file in files:\n",
    "        if file.endswith(\".tar.gz\") or file.endswith(\".tar.bz2\") or file.endswith(\".tar.xz\"):\n",
    "            tar_files.append(os.path.join(root, file))\n",
    "\n",
    "#Unzip the files with different compression methods\n",
    "for tar_file in tar_files:\n",
    "    if tar_file.endswith(\".tar.gz\"):\n",
    "        mode = 'r:gz'\n",
    "    elif tar_file.endswith(\".tar.bz2\"):\n",
    "        mode = 'r:bz2'\n",
    "    elif tar_file.endswith(\".tar.xz\"):\n",
    "        mode = 'r:xz'\n",
    "    else:\n",
    "        mode = 'r'\n",
    "    \n",
    "    try:\n",
    "        with tarfile.open(tar_file, mode) as tar_ref:\n",
    "            extract_dir = os.path.splitext(os.path.splitext(tar_file)[0])[0]\n",
    "            tar_ref.extractall(extract_dir)\n",
    "            print(f\"Extracted {tar_file} to {extract_dir}\")\n",
    "    except (tarfile.ReadError, tarfile.CompressionError) as e:\n",
    "        print(f\"Failed to extract {tar_file}: {e}\")\n",
    "\n",
    "print(\"Extraction process completed.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
